### PR TITLE
support jwt token authen using remote cluster

### PR DESCRIPTION
**Pull request by irisdingbj**
drop an initial version for https://github.com/istio/istio/issues/22918 to get comments.

### CODE REVIEW

File: security/pkg/server/ca/authenticate/kube_jwt.go [hunk 1]
```diff
@@ -83,3 +106,57 @@ func (a *KubeJWTAuthenticator) Authenticate(ctx context.Context) (*Caller, error
...
 }
+
+func (a *KubeJWTAuthenticator) validateRemoteK8sJwt(targetJWT string) ([]string, error) {
+	tokenReview := &v1.TokenReview{}
+	if a.jwtPolicy == jwt.JWTPolicyThirdPartyJWT {
+		tokenReview.APIVersion = "authentication.k8s.io/v1"
+		tokenReview.Kind = "TokenReview"
+		tokenReview.Spec = v1.TokenReviewSpec{
+			Token:     targetJWT,
+			Audiences: []string{"istio-ca"},
```
Comment (linsun) [inline security/pkg/server/ca/authenticate/kube_jwt.go:117]: @myidpt can you confirm audience is always istio-ca?

File: security/pkg/server/ca/authenticate/kube_jwt.go [hunk 2]
```diff
@@ -83,3 +106,57 @@ func (a *KubeJWTAuthenticator) Authenticate(ctx context.Context) (*Caller, error
...
+		}
+		inServiceAccountGroup := false
+		for _, group := range reviewRes.Status.User.Groups {
+			if group == "system:serviceaccounts" {
+				inServiceAccountGroup = true
+				break
+			}
+		}
+		if !inServiceAccountGroup {
+			log.Warnf("the token is not a service account")
```
Comment (linsun) [inline security/pkg/server/ca/authenticate/kube_jwt.go:147]: would be helpful to log found the JWT token in which cluster, however the token is not associated with a service account.

**Comment by linsun** [comment]
@irisdingbj thank you for putting this together, much appreciated!  At the high level it matches what I was thinking.

@myidpt @hzxuzhonghu would appreciate your input as well!

File: security/pkg/server/ca/authenticate/kube_jwt.go (hunk 1 above)
Comment (myidpt) [reply security/pkg/server/ca/authenticate/kube_jwt.go:117]: Yes, that's correct.

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go [hunk 3]
```diff
@@ -149,3 +149,15 @@ func (m *Multicluster) updateHandler() {
 		m.XDSUpdater.ConfigUpdate(req)
 	}
 }
+
+func (m *Multicluster) GetRemoteKubeControllers() map[string]kubernetes.Interface {
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/multicluster.go:153]: I think it is GetRemoteKubeClients

File: security/pkg/server/ca/authenticate/kube_jwt.go [hunk 4]
```diff
@@ -38,10 +47,12 @@ type KubeJWTAuthenticator struct {
 	client      tokenReviewClient
 	trustDomain string
 	jwtPolicy   string
+	mc          *kubecontroller.Multicluster
 }
 
 // NewKubeJWTAuthenticator creates a new kubeJWTAuthenticator.
-func NewKubeJWTAuthenticator(k8sAPIServerURL, caCertPath, jwtPath, trustDomain, jwtPolicy string) (*KubeJWTAuthenticator, error) {
+func NewKubeJWTAuthenticator(mc *kubecontroller.Multicluster, k8sAPIServerURL, caCertPath, jwtPath,
```
Comment (hzxuzhonghu) [inline security/pkg/server/ca/authenticate/kube_jwt.go:54]: nit: only pass in the function to get remote kube clients

File: security/pkg/server/ca/authenticate/kube_jwt.go (hunk 1 above)
Comment (hzxuzhonghu) [reply security/pkg/server/ca/authenticate/kube_jwt.go:117]: donot hardcoded

File: security/pkg/server/ca/authenticate/kube_jwt.go [hunk 5]
```diff
@@ -83,3 +106,57 @@ func (a *KubeJWTAuthenticator) Authenticate(ctx context.Context) (*Caller, error
 		Identities: []string{fmt.Sprintf(identityTemplate, a.trustDomain, callerNamespace, callerServiceAccount)},
 	}, nil
 }
+
+func (a *KubeJWTAuthenticator) validateRemoteK8sJwt(targetJWT string) ([]string, error) {
```
Comment (hzxuzhonghu) [inline security/pkg/server/ca/authenticate/kube_jwt.go:110]: should move to k8sauthn.go

File: security/pkg/server/ca/authenticate/kube_jwt.go [hunk 6]
```diff
@@ -71,7 +83,18 @@ func (a *KubeJWTAuthenticator) Authenticate(ctx context.Context) (*Caller, error
...
-		return nil, fmt.Errorf("failed to validate the JWT: %v", err)
+		// try to validate using remote cluster
+		id, err = a.validateRemoteK8sJwt(targetJWT)
+		if err != nil {
+			return nil, fmt.Errorf("failed to validate the JWT: %v", err)
+		}
+	}
+	if len(id) != 2 {
+		// try to validate using remote cluster
+		id, err = a.validateRemoteK8sJwt(targetJWT)
```
Comment (hzxuzhonghu) [inline security/pkg/server/ca/authenticate/kube_jwt.go:94]: duplicate, merge with the above

File: security/pkg/server/ca/authenticate/kube_jwt.go [hunk 7]
```diff
@@ -71,7 +83,18 @@ func (a *KubeJWTAuthenticator) Authenticate(ctx context.Context) (*Caller, error
 	}
 	id, err := a.client.ValidateK8sJwt(targetJWT, a.jwtPolicy)
 	if err != nil {
-		return nil, fmt.Errorf("failed to validate the JWT: %v", err)
+		// try to validate using remote cluster
+		id, err = a.validateRemoteK8sJwt(targetJWT)
```
Comment (hzxuzhonghu) [inline security/pkg/server/ca/authenticate/kube_jwt.go:87]: @howardjohn Is there a way to know the csr is from a specific cluster? Then we can get rid of this trying all logic

File: security/pkg/server/ca/authenticate/kube_jwt.go (hunk 5 above)
Comment (hzxuzhonghu) [reply security/pkg/server/ca/authenticate/kube_jwt.go:110]: Also reuse one function

**Comment by hzxuzhonghu** [comment]
@lei-tang I see the initial token review is committed by you, why didn't you make use of native tokenreview client directly? Just to confirm if there is some missing point

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go (hunk 3 above)
Comment (irisdingbj) [reply pilot/pkg/serviceregistry/kube/controller/multicluster.go:153]: fixed

File: security/pkg/server/ca/authenticate/kube_jwt.go (hunk 2 above)
Comment (irisdingbj) [reply security/pkg/server/ca/authenticate/kube_jwt.go:147]: fixed

File: security/pkg/server/ca/authenticate/kube_jwt.go (hunk 6 above)
Comment (irisdingbj) [reply security/pkg/server/ca/authenticate/kube_jwt.go:94]: fixed

File: security/pkg/k8s/tokenreview/k8sauthn.go [hunk 8]
```diff
@@ -196,44 +234,52 @@ func (authn *K8sSvcAcctAuthn) ValidateK8sJwt(targetToken, jwtPolicy string) ([]s
...
-
-	payloadBytes, err := base64.RawStdEncoding.DecodeString(payload)
-	if err != nil {
-		return false, fmt.Errorf("failed to decode jwt: %v", err.Error())
+//ValidateRemoteK8sJwt validates aaainst a remote k8s JWT at API server.
+func ValidateRemoteK8sJwt(targetJWT, jwtPolicy string, mc *kubecontroller.Multicluster) ([]string, error) {
+	tokenReview := &k8sauth.TokenReview{}
+	if jwtPolicy == jwt.JWTPolicyThirdPartyJWT {
+		tokenReview.APIVersion = "authentication.k8s.io/v1"
+		tokenReview.Kind = "TokenReview"
```
Comment (hzxuzhonghu) [inline security/pkg/k8s/tokenreview/k8sauthn.go:257]: nit: Using native k8s api, the kind and apiversion is unneccessary

File: security/pkg/k8s/tokenreview/k8sauthn.go [hunk 9]
```diff
@@ -196,44 +234,52 @@ func (authn *K8sSvcAcctAuthn) ValidateK8sJwt(targetToken, jwtPolicy string) ([]s
...
-	jwtSplit := strings.Split(jwt, ".")
-	if len(jwtSplit) != 3 {
-		return false, fmt.Errorf("jwt may be invalid: %s", jwt)
-	}
-	payload := jwtSplit[1]
-
-	payloadBytes, err := base64.RawStdEncoding.DecodeString(payload)
-	if err != nil {
-		return false, fmt.Errorf("failed to decode jwt: %v", err.Error())
+//ValidateRemoteK8sJwt validates aaainst a remote k8s JWT at API server.
```
Comment (hzxuzhonghu) [inline security/pkg/k8s/tokenreview/k8sauthn.go:252]: aaainst

File: security/pkg/k8s/tokenreview/k8sauthn.go [hunk 10]
```diff
@@ -196,44 +234,52 @@ func (authn *K8sSvcAcctAuthn) ValidateK8sJwt(targetToken, jwtPolicy string) ([]s
...
+func ValidateRemoteK8sJwt(targetJWT, jwtPolicy string, mc *kubecontroller.Multicluster) ([]string, error) {
```
Comment (hzxuzhonghu) [inline security/pkg/k8s/tokenreview/k8sauthn.go:253]: Further step, maybe should merge into `ValidateK8sJwt`

**Review by hzxuzhonghu — COMMENTED** [review]
Just some nits, lgtm in genreal

File: security/pkg/k8s/tokenreview/k8sauthn_test.go [hunk 11]
```diff
@@ -120,7 +120,7 @@ func TestOnMockAPIServer(t *testing.T) {
 		authn := NewK8sSvcAcctAuthn(s.httpServer.URL+"/"+tc.cliConfig.reviewPath, tc.cliConfig.tlsCert,
 			tc.cliConfig.reviewerToken)
 
-		_, err := authn.ValidateK8sJwt(tc.cliConfig.jwt, jwt.JWTPolicyThirdPartyJWT)
+		_, err := authn.ValidateK8sJwt(tc.cliConfig.jwt, jwt.JWTPolicyThirdPartyJWT, "kubernetes")
```
Comment (lei-tang) [inline security/pkg/k8s/tokenreview/k8sauthn_test.go:123]: Why the clusterID here in the ValidateK8sJwt() is "kubernetes"? Using "kubernetes" as an ID is confusing and hard to understand. IMO, a cluserID is something like "cluster1", "cluster2", etc.

**Comment by lei-tang** [comment]
> @lei-tang I see the initial token review is committed by you, why didn't you make use of native tokenreview client directly? Just to confirm if there is some missing point

@hzxuzhonghu Can you provide more details of "native tokenreview client"?

File: security/pkg/server/ca/server_test.go [hunk 12]
```diff
@@ -459,7 +459,7 @@ func TestRun(t *testing.T) {
 			tc.expectedAuthenticatorsLen++
 		}
 		server, err := New(tc.ca, time.Hour, false, tc.hostname, tc.port, "testdomain.com", true,
-			jwt.JWTPolicyThirdPartyJWT)
+			jwt.JWTPolicyThirdPartyJWT, "kubernetes")
```
Comment (lei-tang) [inline security/pkg/server/ca/server_test.go:462]: Like a previous comment, using "kubernetes" as an ID is confusing.

File: security/pkg/server/ca/server_test.go [hunk 13]
```diff
@@ -531,7 +531,7 @@ func TestGetServerCertificate(t *testing.T) {
 	}
 
 	server, err := New(ca, time.Hour, false, []string{"localhost"}, 0,
-		"testdomain.com", true, jwt.JWTPolicyThirdPartyJWT)
+		"testdomain.com", true, jwt.JWTPolicyThirdPartyJWT, "kubernetes")
```
Comment (lei-tang) [inline security/pkg/server/ca/server_test.go:534]: Like a previous comment, using "kubernetes" as an ID is confusing.

File: security/pkg/k8s/tokenreview/k8sauthn.go [hunk 14]
```diff
@@ -132,7 +137,7 @@ func (authn *K8sSvcAcctAuthn) reviewServiceAccountAtK8sAPIServer(targetToken, jw
 // Otherwise, return the error.
 // targetToken: the JWT of the K8s service account to be reviewed
 // jwtPolicy: the policy for validating JWT.
-func (authn *K8sSvcAcctAuthn) ValidateK8sJwt(targetToken, jwtPolicy string) ([]string, error) {
+func (authn *K8sSvcAcctAuthn) ValidateK8sJwt(targetToken, jwtPolicy, clusterID string) ([]string, error) {
```
Comment (lei-tang) [inline security/pkg/k8s/tokenreview/k8sauthn.go:140]: The clusterID parameter in ValidateK8sJwt() is passed to getTokenReviewResult(), which only uses clusterID in the logging statements. Is clusterID only used in logging statements? If not, can you point out the places that clusterID impacts the k8s jwt authentication?

Comment (irisdingbj) [reply security/pkg/k8s/tokenreview/k8sauthn.go:140]: yes, this is only used on logging statement.

File: security/pkg/server/ca/server_test.go (hunk 13 above)
Comment (irisdingbj) [reply security/pkg/server/ca/server_test.go:534]: I am using `kubernetes` as default value in testing to align with what we have in runtime:
```   - name: ISTIO_META_CLUSTER_ID
              value: "{{ .Values.global.multiCluster.clusterName | default `Kubernetes` }}"```

File: security/pkg/k8s/tokenreview/k8sauthn_test.go (hunk 11 above)
Comment (irisdingbj) [reply security/pkg/k8s/tokenreview/k8sauthn_test.go:123]: I am using kubernetes as default value in testing to align with what we have in runtime:
```- name: ISTIO_META_CLUSTER_ID
  value: "{{ .Values.global.multiCluster.clusterName | default `Kubernetes` }}"```

File: security/pkg/k8s/tokenreview/k8sauthn.go (hunk 10 above)
Comment (irisdingbj) [reply security/pkg/k8s/tokenreview/k8sauthn.go:253]: OK, we can address this later to merge them together.

File: security/pkg/k8s/tokenreview/k8sauthn.go (hunk 9 above)
Comment (irisdingbj) [reply security/pkg/k8s/tokenreview/k8sauthn.go:252]: fixed

File: security/pkg/k8s/tokenreview/k8sauthn.go (hunk 8 above)
Comment (irisdingbj) [reply security/pkg/k8s/tokenreview/k8sauthn.go:257]: thanks, Fixed~

**Comment by hzxuzhonghu** [comment]
@lei-tang I mean the client-go provided one instead of building it manually.

```
// Create takes the representation of a tokenReview and creates it.  Returns the server's representation of the tokenReview, and an error, if there is any.
func (c *tokenReviews) Create(ctx context.Context, tokenReview *v1.TokenReview, opts metav1.CreateOptions) (result *v1.TokenReview, err error) {
	result = &v1.TokenReview{}
	err = c.client.Post().
		Resource("tokenreviews").
		VersionedParams(&opts, scheme.ParameterCodec).
		Body(tokenReview).
		Do(ctx).
		Into(result)
	return
}
```

File: security/pkg/k8s/tokenreview/k8sauthn.go [hunk 15]
```diff
@@ -162,8 +167,41 @@ func (authn *K8sSvcAcctAuthn) ValidateK8sJwt(targetToken, jwtPolicy string) ([]s
 	if err != nil {
 		return nil, fmt.Errorf("unmarshal response body returns an error: %v", err)
 	}
+	return getTokenReviewResult(tokenReview, clusterID)
+}
+
+// isTrustworthyJwt checks if a jwt is a trustworthy jwt type.
+func isTrustworthyJwt(jwt string) (bool, error) {
```
Comment (hzxuzhonghu) [inline security/pkg/k8s/tokenreview/k8sauthn.go:174]: I think this function is unchanged, not sure why github show it is

File: security/pkg/k8s/tokenreview/k8sauthn.go [hunk 16]
```diff
@@ -196,44 +234,48 @@ func (authn *K8sSvcAcctAuthn) ValidateK8sJwt(targetToken, jwtPolicy string) ([]s
...
 	}
-
-	structuredPayload := &trustWorthyJwtPayload{}
-	err = json.Unmarshal(payloadBytes, &structuredPayload)
-	if err != nil {
-		return false, fmt.Errorf("failed to unmarshal jwt: %v", err.Error())
+	remoteClients := mc.GetRemoteKubeClients()
+	for k, v := range remoteClients {
+		reviewRes, err := v.AuthenticationV1().TokenReviews().Create(context.TODO(), tokenReview, metav1.CreateOptions{})
+		if err != nil {
```
Comment (hzxuzhonghu) [inline security/pkg/k8s/tokenreview/k8sauthn.go:270]: I think when error occurs, we should not do retry in this cluster.

File: security/pkg/k8s/tokenreview/k8sauthn.go (hunk 15 above)
Comment (irisdingbj) [reply security/pkg/k8s/tokenreview/k8sauthn.go:174]: yes, unchanged

File: security/pkg/server/ca/authenticate/kube_jwt.go [hunk 17]
```diff
@@ -69,9 +76,13 @@ func (a *KubeJWTAuthenticator) Authenticate(ctx context.Context) (*Caller, error
 	if err != nil {
 		return nil, fmt.Errorf("target JWT extraction error: %v", err)
 	}
-	id, err := a.client.ValidateK8sJwt(targetJWT, a.jwtPolicy)
-	if err != nil {
-		return nil, fmt.Errorf("failed to validate the JWT: %v", err)
+	id, err := a.client.ValidateK8sJwt(targetJWT, a.jwtPolicy, a.clusterID)
```
Comment (hzxuzhonghu) [inline security/pkg/server/ca/authenticate/kube_jwt.go:79]: same here, we should distinguish if the error is network/ratelimit failure or it is unexpected result that the token not belong to this cluster.

Only fail over to another cluster in the 2nd case

File: security/pkg/k8s/tokenreview/k8sauthn.go (hunk 16 above)
Comment (irisdingbj) [reply security/pkg/k8s/tokenreview/k8sauthn.go:270]: will `apierrors.IsInternalError(err) || apierrors.IsServerTimeout(err) || apierrors.IsTimeout(err)` cover all the cases?

**Comment by lei-tang** [comment]
> @lei-tang I mean the client-go provided one instead of building it manually.
> 
> ```
> // Create takes the representation of a tokenReview and creates it.  Returns the server's representation of the tokenReview, and an error, if there is any.
> func (c *tokenReviews) Create(ctx context.Context, tokenReview *v1.TokenReview, opts metav1.CreateOptions) (result *v1.TokenReview, err error) {
> 	result = &v1.TokenReview{}
> 	err = c.client.Post().
> 		Resource("tokenreviews").
> 		VersionedParams(&opts, scheme.ParameterCodec).
> 		Body(tokenReview).
> 		Do(ctx).
> 		Into(result)
> 	return
> }
> ```

@hzxuzhonghu Thanks for the clarification. The original implementation was in 2018; at that time, the client-go provided one (i.e., func (c *tokenReviews) Create()) may or may not exisit yet. But functionally speaking, they are equivalent.

**Comment by hzxuzhonghu** [comment]
OK, let's do the optimization in a separate pr.

**Review by hzxuzhonghu — APPROVED** [review]
Talked with @irisdingbj online, this is keep consistent with the existing logic. Need to see if feasible to take ClusterID from clientside.

File: security/pkg/k8s/tokenreview/k8sauthn_test.go (hunk 11 above)
Comment (linsun) [reply security/pkg/k8s/tokenreview/k8sauthn_test.go:123]: That is what we currently have all over in multi cluster docs.  If there is a desire to change it, we need to have a separate issue to track, but let's not do it in this PR.

**Review by linsun — APPROVED** [review]
lgtm thank you @irisdingbj

**Comment by linsun** [comment]
@lei-tang @myidpt can you please take a look and approve?  We would want to get this in before code freeze to support our scenario #22934

File: security/pkg/k8s/tokenreview/k8sauthn.go [hunk 18]
```diff
@@ -196,44 +234,48 @@ func (authn *K8sSvcAcctAuthn) ValidateK8sJwt(targetToken, jwtPolicy string) ([]s
...
+			log.Warnf("failed to validate the JWT against cluster %q: due to %v", k, err)
```
Comment (howardjohn) [inline security/pkg/k8s/tokenreview/k8sauthn.go:271]: nit:
```suggestion
			log.Warnf("failed to validate the JWT against cluster %q: %v", k, err)
```

**Review by howardjohn — COMMENTED** [review]
lgtm

File: security/pkg/server/ca/authenticate/kube_jwt.go [hunk 19]
```diff
@@ -69,9 +76,13 @@ func (a *KubeJWTAuthenticator) Authenticate(ctx context.Context) (*Caller, error
...
+	if err != nil || len(id) != 2 {
```
Comment (myidpt) [inline security/pkg/server/ca/authenticate/kube_jwt.go:80]: Falling back to a remote cluster for the verification is fine for the first version of implementation.
I think in the future we can inspect the issuer of the JWT and directly verify it against the mapped cluster. This requires us to maintain a mapping between the clusters and the issuer IDs.

File: security/pkg/k8s/tokenreview/k8sauthn.go (hunk 16 above)
Comment (myidpt) [reply security/pkg/k8s/tokenreview/k8sauthn.go:270]: I think we should do retries in these cases. The only case that we should stop retrying is when the cluster is returning a validation error (which means the JWT isn't signed by itself). And the validation error may be specific on 1) the JWT is not signed by me 2) the JWT is signed by me, but the service account is removed. If we can distinguish them based on the error message, we should continue to next remoteClient when (1) happens, and we should break this loop when (2) happens.

When we have deterministic mapping between the JWT and the clusters. We can do better retries here and greatly reduce the overhead due to the errors.

Does this make sense?

File: security/pkg/server/ca/authenticate/kube_jwt.go (hunk 19 above)
Comment (irisdingbj) [reply security/pkg/server/ca/authenticate/kube_jwt.go:80]: Thanks, I will address this in a following up PR

File: security/pkg/k8s/tokenreview/k8sauthn.go (hunk 18 above)
Comment (irisdingbj) [reply security/pkg/k8s/tokenreview/k8sauthn.go:271]: fixed

**Comment by irisdingbj** [comment]
@howardjohn @myidpt Thanks for the review, PTAL, thanks!

File: security/pkg/k8s/tokenreview/k8sauthn.go (hunk 16 above)
Comment (irisdingbj) [reply security/pkg/k8s/tokenreview/k8sauthn.go:272]: added retry logic for some errors

**Review by lei-tang — APPROVED** [review]

**Comment by lei-tang** [comment]
@linsun LGTM. I hava approved.

**Review by myidpt — APPROVED** [review]
Thanks!

**Pull request merged**